dali/operators/generic/squeeze.cc

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <vector>
+#include <set>
+
+#include "dali/core/math_util.h"
+#include "dali/core/static_switch.h"
+#include "dali/core/tensor_layout.h"
+#include "dali/core/tensor_shape_print.h"
+#include "dali/operators/generic/squeeze.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+
+DALI_SCHEMA(Squeeze)
+  .DocStr("Collapses the dimensions given as axes or axis_names. "
+    "It's an error to collapse dims that would cause the volume to change "
+    "(we can collapse a non-unit dim if a non-collapsed dim is 0).")
+  .NumInput(1, 2)
+  .NumOutput(1)
+  .InputDox(0, "data", "TensorList", "Data to be squeezed")
+  .InputDox(1, "shape_input", "1D TensorList of integers", "Same as `shape` keyword argument")
+  .PassThrough({{0, 0}})
+  .AllowSequences()
+  .SupportVolumetric()
+  .AddOptionalArg<int>("axes", "", std::vector<int>(), true)
+  .AddOptionalArg("axis_names", "", TensorLayout(""));
+
+template <typename Backend>
+Squeeze<Backend>::Squeeze(const OpSpec &spec)
+    : Reshape<Backend>(spec, typename Reshape<Backend>::BypassInit()) {
+  axes_ = spec.GetRepeatedArgument<int>("axes");
+  std::sort(axes_.begin(), axes_.end());
+  axes_.erase(std::unique(axes_.begin(), axes_.end()), axes_.end());
+
+  axis_names_ = spec.GetArgument<TensorLayout>("axis_names");
+
+  this->use_src_dims_ = true;
+}
+
+template <typename Backend>
+bool Squeeze<Backend>::SetupImpl(std::vector<OutputDesc> &output_desc, const Workspace &ws) {
+  output_desc.resize(1);
+  this->SetOutputType(ws);
+
+  GenerateSrcDims(ws);
+  Reshape<Backend>::CalculateOutputShape(ws);
+
+  output_desc[0].type = *(this->output_type_);
+  output_desc[0].shape = this->output_shape_;
+  // return false, because we don't want the executor to allocate anything
+  // - this operator returns pointer to input memory
+  return false;
+}
+
+template <typename Backend>
+void Squeeze<Backend>::GenerateSrcDims(const Workspace &ws) {
+  auto &in = ws.template InputRef<Backend>(0);
+  const auto &input_shape = in.shape();
+  const int ndim = input_shape.sample_dim();
+  auto in_layout = in.GetLayout();
+  DALI_ENFORCE(in_layout.size() == ndim || in_layout.empty(),
+      make_string("Layout for data has size ",
+      in_layout.size(), " but data has ", ndim, " dimensions."));
+
+  this->src_dims_.clear();
+  std::string out_layout;
```

Comment:
Please use TensorLayout. It'll allow you to do things like:
```
TensorLayout out_layout;
out_layout += 'H';
out_layout += 'W';
...

```

Also, it is implemented with a small vector class that doesn't require dynamic allocation for small sizes.